### PR TITLE
RFC 10: ROAST Schnorr signatures

### PR DESCRIPTION
**Pull request by eth-r**
Using ROAST and Schnorr signatures, we can significantly simplify threshold signing while achieving attributability of misbehaviour.

### CODE REVIEW

File: docs/rfc/rfc-10.adoc [hunk 1]
```diff
@@ -0,0 +1,668 @@
...
+* Let R = s⋅G - e⋅P.
+* Fail if is_infinite(R).
+* Fail if not has_even_y(R).
+* Fail if x(R) ≠ r.
+* Return success iff no failure occurred before reaching this point.
+
+For every valid secret key `sk` and message `m`, `Verify(PubKey(sk),m,Sign(sk,m))` will succeed.
+
+We see that the signature is a concatenation of
+a 32-byte x-coordinate or a curve point R,
```
Comment (beaushinkle) [inline docs/rfc/rfc-10.adoc:88]: is this supposed to be "32-byte x-coordinate and a curve point R"?

File: docs/rfc/rfc-10.adoc [hunk 2]
```diff
@@ -0,0 +1,668 @@
...
+while providing the attributability missing from the current tECDSA implementation.
+
+=== Requirements
+
+The Taproot upgrade for Bitcoin enabled Schnorr signature support in 2021.
+BIP-340 defines these Schnorr signatures,
+and our implementation must conform to it.
+
+Schnorr signatures are incompatible with existing Bitcoin scripts,
+requiring the use of taproot scripts instead.
```
Comment (beaushinkle) [inline docs/rfc/rfc-10.adoc:46]: I think a link to an explainer of taproot scripts would help

**Comment by beaushinkle** [comment]
Left a few comments around hyperlinking extra info and some state stuff

Additionally, I think adding more resources like https://www.youtube.com/watch?v=FVW6Hgt_meg and https://github.com/robot-dreams/roast would be sweet

File: docs/rfc/rfc-10.adoc [hunk 3]
```diff
@@ -0,0 +1,668 @@
...
+* A signature sig: a 64-byte array
+
+The algorithm `Verify(pk, m, sig)` is defined as:
+
+* Let P = lift_x(int(pk)); fail if that fails.
+* Let r = int(sig[0:32]); fail if r ≥ p.
+* Let s = int(sig[32:64]); fail if s ≥ n.
+* Let e = int(hash(BIP0340/challenge, bytes(r) || bytes(P) || m)) mod n.
+* Let R = s⋅G - e⋅P.
+* Fail if is_infinite(R).
```
[the remark this answers is not in the document]
Comment (beaushinkle) [reply docs/rfc/rfc-10.adoc]: ah, found the problem: https://docs.asciidoctor.org/asciidoc/latest/syntax-quick-reference/#text-replacements

[the remark this answers is not in the document]
Comment (beaushinkle) [reply docs/rfc/rfc-10.adoc]: image

can get around it by adding escapes:

```
* Fail if `is_infinite\(R)`.
* Fail if not `has_even_y(R)`.
```

**Comment by beaushinkle** [comment]
I'm good here! I think it would be helpful to have someone from @keep-network/go check this out as well

**Review by mhluongo — APPROVED** [review]
Excellent RFC, and overdue for a merge.

File: docs/rfc/rfc-10.adoc [hunk 4]
```diff
@@ -0,0 +1,1192 @@
...
+
+Input:
+
+* The public key pk: a 32-byte array
+* The message m: a 32-byte array
+* A signature sig: a 64-byte array
+
+The algorithm `Verify(pk, m, sig)` is defined as:
+
+* Let `P = lift_x(int(pk))`; fail if that fails.
```
[the remark this answers is not in the document]
Comment (mhluongo) [reply docs/rfc/rfc-10.adoc:80]: Created a follow up issue

https://github.com/keep-network/tbtc-v2/issues/664

File: docs/rfc/rfc-10.adoc [hunk 5]
```diff
@@ -0,0 +1,1192 @@
...
+refund addresses would have to be bech32m P2TR addresses
+rather than legacy address types.
+
+=== BIP-340
+
+link:https://github.com/bitcoin/bips/blob/master/bip-0340.mediawiki[BIP-340] defines Schnorr signatures for the curve secp256k1,
+already widely used in Bitcoin.
+A signature for message m and public key point P
+is defined as the (point, scalar) pair `(R, s)`
+where `s * G = R + hash(R || P || m) * P`.
```
[the remark this answers is not in the document]
Comment (mhluongo) [reply docs/rfc/rfc-10.adoc:59]: Follow up issue

https://github.com/keep-network/tbtc-v2/issues/665

**Pull request merged**